### PR TITLE
Not CSS needed anymore

### PR DESCRIPTION
**Pull request by timonbandit**
Hey @hummal, 
your software is awesome but i think it miss some things.
I've added the options to work without inline css. It's almost the same performance as before. 
And some (all) dependencies are too old.
And benchmark.js was broken.

### CODE REVIEW

File: lib/helper/localFileServer.js
Comment (hummal) [inline lib/helper/localFileServer.js:13]: Remove console.log

**Review by hummal — CHANGES_REQUESTED** [review]

**Comment by hummal** [comment]
Thanks for your work. Please review the requested changes and update the pr.

File: lib/helper/localFileServer.js
Comment (timonbandit) [reply lib/helper/localFileServer.js:13]: Omg. My bad :-)

**Comment by timonbandit** [comment]
Hey @hummal - Done!

**Review by hummal — CHANGES_REQUESTED** [review]
@timonbandit We need a test for your use case (no css given) 
So pr looks good so far. I only want you to add an additional test to check out the valid response for NO css given.

**Comment by timonbandit** [comment]
@hummal Please check tests. 
Works. At least for me :-)

**Review by hummal — COMMENTED** [review]
Code looks good. Testing tests now

**Review by hummal — APPROVED** [review]

**Review by hummal — CHANGES_REQUESTED** [review]
@timonbandit Test not working when requesting localhost URLs without opening a local server before. 

I used the local files instead. But if you like to use the webserver (which includes testing the connection to urls, which I removed by using the files, so it may be a good idea) you need to setup the local file server in your setup before requesting crittr to localhost. 

You should be able to launch tests of crittr in your pr on your own. So you will see if that works out. If not please leave me a msg

**Comment by timonbandit** [comment]
@hummal Can you help me a little.
I'm not sure I understood some tests.
For example - ` MediaQuery 1024 selectors should be included`
And it's work, but you expect to get 0 elements.
The same for ` MediaQuery 800 selectors should be included`

**Review by hummal — CHANGES_REQUESTED** [review]
/lib/helper/localFileServer.js is already a pre configured server to be used for tests. no need to create a server.js in test folder.

**Comment by hummal** [comment]
About the tests. It counts the missing selectors by comparing the result set of selectors with the array of needed selectors. If one selector of expected selectors is missing it will be added to missingSelectors. 

missingSelectors has to be 0 to be good

**Comment by timonbandit** [comment]
Will try to update a bit latter!

**Comment by timonbandit** [comment]
Unfortunately i had to remove "vendor prefixes` test case.
It doesn't work with `nocss` solution. and i don't want to extract it. To much problems for such small case.

File: test/data/test.css
Comment (hummal) [inline test/data/test.css:133]: Not sure why you removed these

File: test/basis.test.js
Comment (hummal) [inline test/basis.test.js:188]: Why don't you write an additional test case for vendor prefixes and leave it for the NOCSS tests only. So we keep the vendors test like it is but don't use it for the NOCSS

File: test/basis.test.js
Comment (hummal) [inline test/basis.test.js:146]: Not sure about that. I want to be sure that those terms won't excluded in the extraction process.

**Review by hummal — CHANGES_REQUESTED** [review]
Please have a look at my comments and change if possible

File: test/data/test.css
Comment (timonbandit) [reply test/data/test.css:133]: To pass the tests :-)
When puppeteer/chome collect css from a page it ignore "all and" 
and then tests failed.

File: test/basis.test.js
Comment (timonbandit) [reply test/basis.test.js:188]: I'm not good in Jest to be honest :-) For me it's a bit complex now. 
I tried to split tests to no-css and with-css, but it make tests slower. 
If you can help me with this - I'll be happy  :-)

File: test/basis.test.js
Comment (timonbandit) [reply test/basis.test.js:146]: In nocss case, `all and` will be dropped by puppeteer.

File: test/basis.test.js
Comment (hummal) [reply test/basis.test.js:188]: Sorry for delay. Iam at vacation in the moment. Will have a look later.

Comment (timonbandit) [reply test/basis.test.js:188]: No stress.
I can use it anyway from my repo :-)
Have a good one!

**Comment by hummal** [comment]
Sorry for delay. Will have a look the next time

File: lib/classes/Crittr.class.js
Comment (hummal) [inline lib/classes/Crittr.class.js:10]: I don't see why we should replace deepmerge with lodash/merge

File: lib/classes/CssTransformator.class.js
Comment (hummal) [inline lib/classes/CssTransformator.class.js:5]: I don't see why we should replace deepmerge with lodash/merge

File: package.json
Comment (hummal) [inline package.json:2]: I can not merge this file

**Review by hummal — CHANGES_REQUESTED** [review]
PLease review the comments. 
I will have a look at the tests now

File: package.json
Comment (timonbandit) [reply package.json:2]: Don't need to merge it :-)

File: lib/classes/CssTransformator.class.js
Comment (timonbandit) [reply lib/classes/CssTransformator.class.js:5]: Because deepmerge cannot merge puppeteer as external browser. It fails with error.

File: package.json
Comment (hummal) [reply package.json:2]: If you change it that way I can not accept the whole pull request. 
You can not change the whole urls and versions in your pull request. Adding or removing packages is needed sometimes but releasing new versions is a thing of the maintainers not the contributors. 

Do we agree?

File: lib/classes/CssTransformator.class.js
Comment (hummal) [reply lib/classes/CssTransformator.class.js:5]: You should have a look at [Deepmerge:ismergeableobject](https://www.npmjs.com/package/deepmerge#ismergeableobject) which should resolve your problem with puppeteer. 

I don't think it is a good idea to include a lodash merge which mutates Object itself just to exclude instances from being cloned. 

This function will enable you to exclude all instanciated objects in merge. Which is only the puppeteer instance. 

Iam not sure if that works out. You can try it. Iam working on fixing the tests.

**Review by hummal — CHANGES_REQUESTED** [review]
- package.json has to be without url and version changes. That's bad behaviour in open source. Not sure why you did it
- please check the lodash/merge option I wrote in comment. Want to prevent Object from being mutated with lodash

**Comment by hummal** [comment]
Ah I see you already released a new package on npm.

**Comment by timonbandit** [comment]
Hey @hummal Yes. I did the release. Because the current version is impossible to use from npm with puppeteer and no-css.

**Comment by timonbandit** [comment]
Please feel free to close it

**Comment by hummal** [comment]
> Hey @hummal Yes. I did the release. Because the current version is impossible to use from npm with puppeteer and no-css.

You could have used your repository link in package.json for that. 

Iam currently working at your fork to help with the tests. If you reset the package.json and have a look at the lodash merge thingy Iam sure we can release it this weekend

**Comment by timonbandit** [comment]
> 

Unfortunately I have no time this weekend :-(
Just FYI - i tried to make critical generation microservice based on Google Cloud Functions. but seems it's not a good place to run crittr. any page generation took more 30second. 
Localy it works like a charm with external puppeteer. 
1200 pages generation took around 15 min.
But on Google cloud functions 15 min for 100 pages :-)

**Comment by hummal** [comment]
Okay, if you could fix the package.json stuff I will look on the rest.

Microservice: I was thinking of dealing with AWS and crittr. So similar approach. Problem is puppeteer. Too heavy for lambda and co. 
Not sure there is a way arround.

**Comment by timonbandit** [comment]
> Okay, if you could fix the package.json stuff I will look on the rest.
> 
> Microservice: I was thinking of dealing with AWS and crittr. So similar approach. Problem is puppeteer. Too heavy for lambda and co. Not sure there is a way arround.

I did it. Nodejs+express+bullMQ+crittr+Redis. 
Works perfectly. But on local PC or on docker.

**Comment by hummal** [comment]
It is the cloud platform that is not a good match for puppeteer. You can read it in Google Lighthouse blogs where they are talking about that topic.

**Comment by hummal** [comment]
I changed your stuff. have a look if you agree. 

And please test the puppeteer deepmerge problem. I included the isPlainObject function. This may already solve the problem.

**Comment by timonbandit** [comment]
> I changed your stuff. have a look if you agree. I resetted the package.json aswell.
> 
> And please test the puppeteer deepmerge problem. I included the isPlainObject function. This may already solve the problem.

Thank you!
I'll have a look on Monday!

**Comment by hummal** [comment]
@timonbandit did you had time? Wanna proceed ?

**Comment by timonbandit** [comment]
> @timonbandit did you had time? Wanna proceed ?

@hummal Hey! Sorry for the delay! 
Extremely busy week - On my job we're preparing for Black Friday - working without Holidays :-)
I think next week. Feel free to close the PR. I'll open the new one

**Comment by hummal** [comment]
@timonbandit all cool. relax. 
we dont need to open a new one. Everything is fixed here. You only need to check if you can use the browser option with deepmerge.

**Comment by timonbandit** [comment]
@hummal - Updated. Please check if you have time :-)

File: lib/classes/CssTransformator.class.js
Comment (hummal) [inline lib/classes/CssTransformator.class.js:40]: What do you mean?

**Review by hummal — COMMENTED** [review]
Iam worried about the added comment about the correctness of the added lines. Do you have some intel?

deepmerge with your browser instance was working?

File: lib/classes/CssTransformator.class.js
Comment (timonbandit) [reply lib/classes/CssTransformator.class.js:40]: 1st - `this._PSUEDO_SELECTOR_REGEXP` - looks like a typo. `PSUEDO` instead of `PSEUDO`
2nd - i didn't find a place where you use it.

Comment (hummal) [reply lib/classes/CssTransformator.class.js:40]: True. Didn't find it either. Will remove it after merge

**Review by hummal — APPROVED** [review]

**Pull request merged**
